File: todolist/src/types/asset.ts
export interface Asset {
  id: string;
  name: string;
  category: string;
  purchase_date: string;
  price: number;
  currency: string;
  quantity: number;
  brand: string;
  model: string;
  warranty_expiry: string | null;
  status: string;
  condition: string;
  notes: string;
  is_sentimental: boolean;
  origin: string;
  related_people: string;
  related_stories: string;
  retired_at: string | null;
  farewell_message: string;
  created_at: string;
  updated_at: string;
}

export interface CreateAssetParams {
  name: string;
  category?: string;
  purchase_date: string;
  price?: number;
  currency?: string;
  quantity?: number;
  brand?: string;
  model?: string;
  warranty_expiry?: string | null;
  status?: string;
  condition?: string;
  notes?: string;
}

export interface UpdateAssetParams {
  name?: string;
  category?: string;
  purchase_date?: string;
  price?: number;
  currency?: string;
  quantity?: number;
  brand?: string;
  model?: string;
  warranty_expiry?: string | null;
  status?: string;
  condition?: string;
  notes?: string;
}

export interface AssetStats {
  total_value: number;
  total_count: number;
}

/** Preset categories (7) */
export const ASSET_CATEGORIES = [
  { key: 'electronics', label: '电子数码', icon: '💻' },
  { key: 'appliances', label: '家居电器', icon: '🏠' },
  { key: 'books', label: '书籍文具', icon: '📚' },
  { key: 'entertainment', label: '娱乐休闲', icon: '🎮' },
  { key: 'tools', label: '工具设备', icon: '🛠️' },
  { key: 'sports', label: '运动户外', icon: '🏃' },
  { key: 'other', label: '其他', icon: '🎯' },
] as const;

export const ASSET_STATUSES = [
  { key: 'in_use', label: '使用中', color: '#10B981' },
  { key: 'idle', label: '闲置', color: '#F59E0B' },
  { key: 'sold', label: '已出售', color: '#6366F1' },
  { key: 'broken', label: '已损坏', color: '#EF4444' },
  { key: 'lost', label: '已丢失', color: '#9CA3AF' },
] as const;

export const ASSET_CONDITIONS = [
  { key: 'new', label: '全新' },
  { key: 'good', label: '良好' },
  { key: 'fair', label: '一般' },
  { key: 'poor', label: '较差' },
] as const;

export function getCategoryLabel(key: string): string {
  return ASSET_CATEGORIES.find(c => c.key === key)?.label ?? key;
}

export function getCategoryIcon(key: string): string {
  return ASSET_CATEGORIES.find(c => c.key === key)?.icon ?? '🎯';
}

export function getStatusLabel(key: string): string {
  return ASSET_STATUSES.find(s => s.key === key)?.label ?? key;
}

export function getStatusColor(key: string): string {
  return ASSET_STATUSES.find(s => s.key === key)?.color ?? '#9CA3AF';
}

export function getConditionLabel(key: string): string {
  return ASSET_CONDITIONS.find(c => c.key === key)?.label ?? key;
}

/** Calculate warranty remaining days. Returns null if no warranty. Negative = expired. */
export function getWarrantyDays(expiry: string | null): number | null {
  if (!expiry) return null;
  const now = new Date();
  const end = new Date(expiry);
  return Math.ceil((end.getTime() - now.getTime()) / (1000 * 60 * 60 * 24));
}

/** Warranty color based on remaining days */
export function getWarrantyColor(days: number | null): string | undefined {
  if (days === null) return undefined;
  if (days < 0) return '#9CA3AF'; // expired - gray
  if (days <= 30) return '#EF4444'; // expiring soon - red
  if (days <= 90) return '#F59E0B'; // warning - yellow
  return '#10B981'; // fine - green
}

export function formatWarranty(days: number | null): string | null {
  if (days === null) return null;
  if (days < 0) return `已过期 ${Math.abs(days)} 天`;
  if (days === 0) return '今日到期';
  if (days <= 30) return `还有 ${days} 天`;
  if (days <= 90) return `剩 ${days} 天`;
  return `剩余 ${days} 天`;
}

File: todolist/src-tauri/src/models/asset.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Asset {
    pub id: String,
    pub name: String,
    pub category: String,
    pub purchase_date: String,
    pub price: f64,
    pub currency: String,
    pub quantity: i32,
    pub brand: String,
    pub model: String,
    pub warranty_expiry: Option<String>,
    pub status: String,
    pub condition: String,
    pub notes: String,
    pub is_sentimental: bool,
    pub origin: String,
    pub related_people: String,
    pub related_stories: String,
    pub retired_at: Option<String>,
    pub farewell_message: String,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Deserialize)]
pub struct CreateAssetParams {
    pub name: String,
    #[serde(default = "default_category")]
    pub category: String,
    pub purchase_date: String,
    #[serde(default)]
    pub price: f64,
    #[serde(default = "default_currency")]
    pub currency: String,
    #[serde(default = "default_quantity")]
    pub quantity: i32,
    #[serde(default)]
    pub brand: String,
    #[serde(default)]
    pub model: String,
    pub warranty_expiry: Option<String>,
    #[serde(default = "default_status")]
    pub status: String,
    #[serde(default = "default_condition")]
    pub condition: String,
    #[serde(default)]
    pub notes: String,
    #[serde(default)]
    pub is_sentimental: bool,
    #[serde(default)]
    pub origin: String,
    #[serde(default)]
    pub related_people: String,
    #[serde(default)]
    pub related_stories: String,
    pub retired_at: Option<String>,
    #[serde(default)]
    pub farewell_message: String,
}

#[derive(Debug, Deserialize)]
pub struct UpdateAssetParams {
    pub name: Option<String>,
    pub category: Option<String>,
    pub purchase_date: Option<String>,
    pub price: Option<f64>,
    pub currency: Option<String>,
    pub quantity: Option<i32>,
    pub brand: Option<String>,
    pub model: Option<String>,
    pub warranty_expiry: Option<String>,
    pub status: Option<String>,
    pub condition: Option<String>,
    pub notes: Option<String>,
    pub is_sentimental: Option<bool>,
    pub origin: Option<String>,
    pub related_people: Option<String>,
    pub related_stories: Option<String>,
    pub retired_at: Option<String>,
    pub farewell_message: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AssetStats {
    pub total_value: f64,
    pub total_count: i64,
}

fn default_category() -> String {
    "other".into()
}
fn default_currency() -> String {
    "CNY".into()
}
fn default_quantity() -> i32 {
    1
}
fn default_status() -> String {
    "in_use".into()
}
fn default_condition() -> String {
    "good".into()
}
